plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServer.java

```diff
@@ -0,0 +1,215 @@
+
+  /**
+   * Returns a {@link CloudSdkServer} instance associated with the
+   * {@code server} or a new {@link CloudSdkServer} instance if a
+   * {@link CloudSdkServer} instance cannot be found for {@code server}.
+   *
+   * @param server the generic sever
+   * @return a new {@link CloudSdkServer} instance or the one associated with
+   *         {@code server}
+   */
+  public static CloudSdkServer getCloudSdkServer(IServer server) {
+    CloudSdkServer cloudSdkServer = (CloudSdkServer) server.getAdapter(CloudSdkServer.class);
+    if (cloudSdkServer == null) {
+      cloudSdkServer = (CloudSdkServer) server.loadAdapter(CloudSdkServer.class,
+                                                           new NullProgressMonitor());
+    }
+    return cloudSdkServer;
+  }
+
+  @Override
+  public IStatus canModifyModules(IModule[] add, IModule[] remove) {
+    if (add != null) {
+      for (IModule module : add) {
+        if (module.getProject() != null) {
+          IStatus status = FacetUtil.verifyFacets(module.getProject(), getServer());
+          if (status != null && !status.isOK()) {
+            return status;
+          }
+        }
+      }
+    }
+    return Status.OK_STATUS;
+  }
+
+  @Override
+  public IModule[] getChildModules(IModule[] module) {
+    // This is the same logic as in the GaeServer. The GaeServer would be
+    // replaced by the
+    // CloudSdkServer.
+    if (module[0] != null && module[0].getModuleType() != null) {
+      IModule thisModule = module[module.length - 1];
+      IModuleType moduleType = thisModule.getModuleType();
+      if (moduleType != null && "jst.ear".equals(moduleType.getId())) { //$NON-NLS-1$
+        IEnterpriseApplication enterpriseApplication = (IEnterpriseApplication) thisModule.loadAdapter(IEnterpriseApplication.class,
+                                                                                                       null);
+        if (enterpriseApplication != null) {
+          IModule[] earModules = enterpriseApplication.getModules();
+          if (earModules != null) {
+            return earModules;
+          }
+        }
+      } else if (moduleType != null && "jst.web".equals(moduleType.getId())) { //$NON-NLS-1$
+        IWebModule webModule = (IWebModule) thisModule.loadAdapter(IWebModule.class, null);
+        if (webModule != null) {
+          IModule[] modules = webModule.getModules();
+          return modules;
+        }
+      }
+    }
+    return new IModule[0];
+  }
+
+  @Override
+  public IModule[] getRootModules(IModule module) throws CoreException {
+    IStatus status = canModifyModules(new IModule[] { module }, null);
+    if (status != null && !status.isOK()) {
+      throw new CoreException(status);
+    }
+    IModule[] parents = doGetParentModules(module);
+    if (parents.length > 0) {
+      return parents;
+    }
+    return new IModule[] { module };
+  }
+
+  @SuppressWarnings("unchecked")
```

Comment:
@elharo unfortunately the method in the parent class does not declare the generic type parameter